include/oneapi/dpl/pstl/hetero/dpcpp/sycl_traits.h

```diff
@@ -0,0 +1,152 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+// This file contains some specialization SYCL traits for some oneDPL types.
+//
+// Fancy iterators and internal functors which are device copyable when their
+// template arguments are also device copyable should be explicitly specialized
+// as such. This is important when template argument member variables may be
+// device copyable but not trivially copyable.
+// Include this header before a kernel submit SYCL code.
+
+#ifndef _ONEDPL_SYCL_TRAITS_H
+#define _ONEDPL_SYCL_TRAITS_H
+
+#if __INTEL_LLVM_COMPILER && (__INTEL_LLVM_COMPILER < 20240100)
+
+#    define _ONEDPL_DEVICE_COPYABLE(TYPE)                                                                              \
+        template <typename... Ts>                                                                                      \
+        struct sycl::is_device_copyable<TYPE<Ts...>, ::std::enable_if_t<!std::is_trivially_copyable_v<TYPE<Ts...>>>>   \
+            : ::std::conjunction<sycl::is_device_copyable<Ts>...>                                                      \
+        {                                                                                                              \
+        };
+
+#else
+
+#    define _ONEDPL_DEVICE_COPYABLE(TYPE)                                                                              \
+        template <typename... Ts>                                                                                      \
+        struct sycl::is_device_copyable<TYPE<Ts...>> : ::std::conjunction<sycl::is_device_copyable<Ts>...>             \
+        {                                                                                                              \
+        };
+
+#endif
+
+using namespace oneapi::dpl::__internal;
+
+_ONEDPL_DEVICE_COPYABLE(__not_pred)
+_ONEDPL_DEVICE_COPYABLE(__reorder_pred)
+_ONEDPL_DEVICE_COPYABLE(__equal_value_by_pred)
+_ONEDPL_DEVICE_COPYABLE(__equal_value)
+_ONEDPL_DEVICE_COPYABLE(__not_equal_value)
+_ONEDPL_DEVICE_COPYABLE(__transform_functor)
+_ONEDPL_DEVICE_COPYABLE(__transform_if_unary_functor)
+_ONEDPL_DEVICE_COPYABLE(__transform_if_binary_functor)
+_ONEDPL_DEVICE_COPYABLE(__replace_functor)
+_ONEDPL_DEVICE_COPYABLE(__replace_copy_functor)
+_ONEDPL_DEVICE_COPYABLE(zip_forward_iterator)
```

Comment:
It turns out `zip_forward_iterator` is only used on the host side (from `include/oneapi/dpl/pstl/algorithm_impl.h`), within functions which are enabled for host policies.  It is probably not necessary to specialize it as device copyable.